tests/integration_tests/device/cuda/navigation_validation.hpp

```diff
@@ -0,0 +1,304 @@
+ *
+ * Mozilla Public License Version 2.0
+ */
+
+#pragma once
+
+// Project include(s)
+#include "detray/core/detector.hpp"
+#include "detray/detectors/bfield.hpp"
+#include "detray/navigation/detail/ray.hpp"
+#include "detray/propagator/line_stepper.hpp"
+#include "detray/propagator/rk_stepper.hpp"
+#include "detray/test/fixture_base.hpp"
+#include "detray/test/navigation_validation_config.hpp"
+#include "detray/test/utils/detector_scan_utils.hpp"
+#include "detray/test/utils/detector_scanner.hpp"
+#include "detray/test/utils/navigation_validation_utils.hpp"
+#include "detray/tracks/tracks.hpp"
+
+// Vecmem include(s)
+#include <vecmem/memory/cuda/device_memory_resource.hpp>
+#include <vecmem/memory/cuda/managed_memory_resource.hpp>
+#include <vecmem/memory/host_memory_resource.hpp>
+#include <vecmem/memory/memory_resource.hpp>
+#include <vecmem/utils/cuda/copy.hpp>
+
+namespace detray::cuda {
+
+/// Launch the navigation validation kernel
+///
+/// @param[in]
+template <typename bfield_t, typename detector_t,
+          typename intersection_record_t>
+void navigation_validation_device(
+    typename detector_t::view_type det_view,
+    const propagation::config<typename detector_t::scalar_type> &cfg,
+    bfield_t field_data,
+    vecmem::data::jagged_vector_view<
+        typename intersection_record_t::intersection_type> &candidates_data,
+    vecmem::data::jagged_vector_view<const intersection_record_t>
+        &intersection_traces_view,
+    vecmem::data::jagged_vector_view<
+        typename intersection_record_t::intersection_type>
+        &intersection_records_view);
+
+/// Test function for propagator on the device
+template <typename bfield_t, typename detector_t,
+          typename intersection_record_t>
+inline auto run_navigation_validation(
+    vecmem::memory_resource *host_mr, vecmem::memory_resource *dev_mr,
+    detector_t *det,
+    const propagation::config<typename detector_t::scalar_type> &cfg,
+    bfield_t field_data,
+    const std::vector<std::vector<intersection_record_t>>
+        &host_intersection_traces)
+    -> vecmem::jagged_vector<
+        typename intersection_record_t::intersection_type> {
+
+    using intersection_t = typename intersection_record_t::intersection_type;
+
+    // Helper object for performing memory copies (to CUDA devices)
+    vecmem::cuda::copy cuda_cpy;
+
+    // Copy the detector to device and get its view
+    auto det_buffer = detray::get_buffer(*det, *dev_mr, cuda_cpy);
+    auto det_view = detray::get_data(det_buffer);
+
+    // Create vector buffer for intersection traces
+    std::vector<std::size_t> capacities;
+    for (const auto &trace : host_intersection_traces) {
+        capacities.push_back(trace.size());
+    }
+
+    // Create navigator candidates buffer
+    vecmem::cuda::managed_memory_resource mng_mr;
+    printf("Max candidates: %ld\n", det->n_max_candidates());
+    vecmem::copy copy;
+    auto navigation_buffer =
```

Comment:
This is where I try to build the navigation buffer. In this version I tried it using managed memory, but there is also a commented version of a direct copy, which is where I would like to arrive in the end. The buffer is then supposed to be used in the kernel defined in ```navigation_validation.cu```, but it seems to have zero capacity in all of its subvectors. I am sure I am just overlooking something silly here, but could you please have a second look @krasznaa ?

Reply:
Hmm... I'm learning some new stuff about this code myself as well...

First off, the returned buffer is a resizable one. This I didn't realize. Now I understand a little better why `vecmem::copy::setup(...)` is called in the traccc code in such a weird way at the moment.

So... In line 87, after calling `vecmem::copy::setup(...)`, the buffer should have the following setup:
  - `navigation_buffer.size() == host_intersection_traces.size()`;
  - `navigation_buffer.capacity() = host_intersection_traces.size()`;
  - `navigation_buffer[i].size() == 0u`;
  - `navigation_buffer[i].capacity() == det->n_max_candidates()`.

(Note that this is not actual code. Buffers can't be interacted with like this, you'd need to make a `vecmem::jagged_device_vector` around this buffer to investigate its payload like this.)

If that's not what you see in the kernel code (which might be the case as much as I understood you...), then I'm not sure right now where things go wrong. Since the buffer doesn't seem to obviously go out of scope before the kernel would run. 🤔

Reply:
Indeed, the problem is that ```navigation_buffer[i].capacity() == 0``` for some reason, while ```det->n_max_candidates()``` prints 24, which makes sense

Reply:
I initially tried to do this with the two memeory resources and without the copy object, but that showed the same result

Reply:
🤔 Curious. It's good that you tried with separate host and device memory resources as well. Since with SYCL "shared memory" doesn't always work exactly as you'd expect. (Though with CUDA it should...)

As long as right here the capacities are reported correctly (you should probably write a couple of extra lines of code for checking that), then the buffer still must get deleted before the kernel would run. I can just not see in the code yet where that would happen. 😦

Reply:
Putting a ```jagged_device_vector``` around it and reading out the capacities in host code prints the expected "24"

Reply:
OK, no, in the kernel, it is also "24" with the managed memory resource, but in the navigator it still triggers this assertion:
```
assert(intersections.size() < intersections.capacity() &&
                       "Navigation cache size too small");
```
I added a universal reference in the propagation call, let me take a look...

Reply:
One unusual thing that I wanted to mention is how you pass around non-const references to all these view objects. Note that if you do that for the internal vectors of a jagged vector, then you're making a mistake... Since the internal vectors of a jagged vector are created on the spot, as temporary objects.

https://github.com/acts-project/vecmem/blob/main/core/include/vecmem/containers/jagged_device_vector.hpp#L60-L63

(Notice the missing `&` from those types.) Could you end up with an invalid object inside of the device code? 🤔

Reply:
You should try to run your test executable through `compute-sanitizer`. It may be able to catch the memory error that we're after... 🤔

Reply:
Ok, I think with the managed memory resource it works now, but when I use the the device and host memory resource in the commented line below it still fails. Am I missing a copy operation somewhere?

Reply:
Oh, I just saw that the copy operation is hidden in the track finder and fitter. So, yes, I definitely missed that step

Reply:
Alright, looks good now! Turns out I missed the actual copy to device, my bad! When the code complained it could not call ```clear()``` on a resizable vector of zero capacity, I commented that line, which lead to a valid instance of the navigation buffer going above its capacity when I used the managed memory resource later. Hence, me thinking it did not work with managed memory either. Thanks for having a second look!